### ISSUE TITLE
Web Mercator (EPSG:3857) not detected correctly

### ISSUE THREAD

**Issue by mbloch** [matafokka/geotiff-geokeys-to-proj4#7]
I'm attaching a sample DEM file in webmercator coordinates. `geotiff-geokeys-to-proj4` gives a Proj4 string for an ellipsoid, which is incorrect: 
```
+proj=merc +lat_0=0 +lon_0=0 +x_0=0 +y_0=0 +ellps=WGS84 +a=6378137 +b=6356752.314245179 +pm=0 +axis=en +no_defs
```

`gdalinfo` gives a correct string:
```
+proj=merc +a=6378137 +b=6378137 +lat_ts=0 +lon_0=0 +x_0=0 +y_0=0 +k=1 +units=m +nadgrids=@null +wktext +no_defs
```

These are the geokeys:
```json
{
  "GTModelTypeGeoKey": 1,
  "GTRasterTypeGeoKey": 1,
  "GTCitationGeoKey": "WGS 84 / Pseudo-Mercator",
  "GeogCitationGeoKey": "WGS 84",
  "GeogAngularUnitsGeoKey": 9102,
  "ProjectedCSTypeGeoKey": 3857,
  "ProjLinearUnitsGeoKey": 9001
}
```

And here is the file:

[dem.tif.zip](https://github.com/matafokka/geotiff-geokeys-to-proj4/files/14913099/dem.tif.zip)

**Comment by matafokka** [comment]
Hello, thanks for reporting this! I'll be able to check this at the end of this week. Wanted to do that today but forgot that I have unfinished work for this project to do. I'm sorry for the delay.

**Comment by mbloch** [comment]
Thanks for looking! I haven't had time to look at the source code to find a possible fix... I did look at the GeoTIFF spec and noticed that `ProjectedCSTypeGeoKey` belongs to GeoJSON version `1.0`, and has been replaced by `ProjectedCRSGeoKey`...  I wonder if the key change is related to this issue.

**Comment by matafokka** [comment]
> I wonder if the key change is related to this issue

Good call! I've recently introduced GeoTIFF 1.1 support, maybe alias is broken. There's also wrong ellipsoid axis, and `+lat_0` being used instead of `+lat_ts`. I can't tell right away what's wrong, need to look into EPSG database.

**Comment by matafokka** [comment]
Hello! I've got some information regarding ellipsoid axes. I've manually traced everything down to ellipsoid, here's my steps:

1. CRS code is 3857 as you can see in the geokeys.
2. This CRS doesn't have a datum, but it has a base CRS with 4326 code:

![image](https://github.com/matafokka/geotiff-geokeys-to-proj4/assets/34414488/7cffdc91-5e6c-45e8-8ece-b63d55b887f1)

3. The base CRS has a datum with 6326 code:

![image](https://github.com/matafokka/geotiff-geokeys-to-proj4/assets/34414488/1660048f-dc2b-4123-ad56-916c6cde0944)

4. This datum is actually a datum ensemble:

![image](https://github.com/matafokka/geotiff-geokeys-to-proj4/assets/34414488/af360bc2-96e5-4f91-bd90-32e8a355304e)

5. This ensemble has 7 members:

![image](https://github.com/matafokka/geotiff-geokeys-to-proj4/assets/34414488/123cc815-1734-45c0-9275-f272b6d26eae)

6. My library takes one with the greatest datum_sequence field. Just in case I checked all 7 members, all of them have an ellipsoid with 7030 code.

![image](https://github.com/matafokka/geotiff-geokeys-to-proj4/assets/34414488/8e2cec02-f9ca-47be-a6a0-ed01263892c3)

7. This ellipsoid has semi-major axis and inverse flattening set:

![image](https://github.com/matafokka/geotiff-geokeys-to-proj4/assets/34414488/2a03c542-fbce-4917-8b53-83ae3a730039)

8. So there you have it, two different axes.

Moreover, GDAL will give you WKT with a semi-major axes and inverse flattening as well, but proj4 string is indeed like yours:

![image](https://github.com/matafokka/geotiff-geokeys-to-proj4/assets/34414488/2827f898-c837-484a-ac5e-5e5e8bd934d3)

Since I'm using a EPSG database (geokeys are just EPSG codes, so database from epsg.org is one true data source), and GDAL itself gives the same ellipsoid in WKT, the only sane conclusion is that GDAL's proj4 string building algorithm produces incorrect ellipsoid.

___

I still have to check why `+lat_0` is used instead of `+lat_ts`, so stay tuned :)

**Comment by matafokka** [comment]
> I still have to check why +lat_0 is used instead of +lat_ts, so stay tuned :)

Looks like a different name for the same parameter, added an override. Please try new 2024.4.12 release which should be available in like 10 minutes.

I guess this issue is resolved, but if you have another opinion, feel free to reopen it. Thanks for the report!

**Comment by matafokka** [comment]
Huh, in both QGIS and my viewer, GDAL's proj4 string yields right results. I even tried converting to another format, and QGIS still ignores inverse flattening. I'll check against another GIS, but still, am I missing something?

**Comment by mbloch** [comment]
I see your point, it seems kind of crazy to translate the WGS84 ellpsoid to spherical axes (+a=6378137 +b=6378137) in the Proj4 string. But that's what you have to do if you use `+proj=merc` for your projection name. If you pair `merc` with ellipsoidal axes, like your library is doing, then the resulting Proj4 string will produce incorrect results. That's because Proj4's `merc` is not the same as "Popular Visualisation Pseudo-Mercator."

The form that GDAL produces is in common use, and it works. It's what you find on the EPSG website (https://epsg.io/3857). 

There's another Proj.4 string for Web Mercator that is more correct-seeming, and it's what the Proj website currently gives you: (https://proj.org/en/9.4/operations/projections/webmerc.html). This form is
`+proj=webmerc +datum=WGS84`.  Here, `webmerc` is the equivalent of WKT's "Popular Visualisation Pseudo-Mercator." It's a more literal translation for the WKT string that you find in the EPSG database.

To sum up, there are two valid Proj4 strings for Web Mercator/EPSG:3857. One is closer to the WKT form, the other is in common use and is compatible with more software. 

If you want to maintain compatibility with `proj4js`, I suggest that you treat EPSG:3857 as a special case and use the same string as GDAL. Or, you could switch to `webmerc` to give a more correct-seeming string, and lose backwards compatibility with software that hasn't implemented `webmerc`. I would vote for the first option.

**Comment by matafokka** [comment]
Ah, I see, thank you! I have mappings of EPSG conversion codes to `+proj` parameters. I don't remember what my exact sources were, but it seems to work in most cases. For cases like yours, I'll just create exceptions.

Do you happen to know other such projections? I'd happily populate the list tomorrow :)

**Comment by matafokka** [comment]
I've added an override which will set a sphere even if geokeys define an ellipsoid. While normally geokeys will override original CRS, I believe, we should make an exception for Web Mercator.

Please update to the 2024.4.13 release.

**Comment by mbloch** [comment]
Thanks for this update :) I'm not aware of other special cases relating to WKT <-> Proj4 translation, but I'll keep my eyes out for others.

**Comment by matafokka** [comment]
Thank you for your help, appreciate the effort :) It warms my heart knowing there's less potential problems :D
